### ISSUE TITLE
Feature Request: Play Next Folder in Directory

### ISSUE THREAD

**Issue by Level99OCR** [droidmjt/Droidsound#182]
Long time, MJT! Been using the latest beta on Dropbox since it came out, seems pretty stable in all aspects (I have a single bug report to file after this feature request opening).

I thoroughly enjoy the play next file in directory feature in Droidsound E. This works for stuff like GBS and NSF, meaning I can have multiple games in one directory and have it go from one to the next without intervening or going back to the filebrowser.

Would it be possible to implement the same feature on a folder-level scale? Example: I have multiple subfolders or zips in a main folder, and when one zip/subfolder is done playing I would like it to automatically go to the next subfolder in that main folder.

**Comment by droidmjt** [comment]
I understand what you mean, I need to do some experimenting with this.

**Comment by droidmjt** [comment]
this feature is doable but might have nasty consequences, for example, if you have 30 zips in /mnt/sdcard/Music folder, this feature would create a playqueue of all the contents found inside those 30 zips. Making it advance to one zip at a time would require quite a lot changes :(

I'll see further how it goes with this.

**Comment by Level99OCR** [comment]
I gotcha, I didn't realize that the playqueue pre-loads everything so the memory issue makes sense.

Maybe a solution would be to add a feature where an entire subfolder could be added to a playlist or favorites. Do playlist and favorites handle playqueues differently?

**Comment by droidmjt** [comment]
>>Do playlist and favorites handle playqueues differently?
not really, same deal. 

Currently I have other tasks, I'll get back to this once I'm done with the other things. Maybe there is another kind of approach.

**Comment by droidmjt** [comment]
there is files-only mode in playlist settings, enable that and see what happens if you queue 100000 songs :)

**Comment by Level99OCR** [comment]
Testing it now. Getting a weird bug where trying to type in the name to a new playlist, every other (alternating) character I enter blanks the line. Entering Genesis:
G -> *blank* -> n -> *blank* -> s -> *blank* -> s 
Playlist is then saved as just "s". 

I have a playlist of just about 1500 testing right now. Order is a tiny bit random. The sorts are fine based on the criteria used but it doesn't seem to separate the sorts based on sub-directory or zip file name. Games with multiple songs get played out of order. It would be nice to have those groupings together but it's not a dealbreaker in my opinion. What would be nice is for, on the playing file info page, the name of the Game were also presented. Since the playlists may result in semi-random orderings right now, it's hard to distinguish which game they're from based on the song title alone.

It's definitely a step in the right direction, and I'll be using it to make work playlists that are more convenient than the other manual workarounds I was doing! :)

**Comment by droidmjt** [comment]
i'll test and fix the playlist creation at once...wth

**Comment by droidmjt** [comment]
in fact, type here how you want the ordering to be done, let me know if you need the fields names used in SQL

**Comment by droidmjt** [comment]
How to reproduce the sorting problem the fastest way?

**Comment by Level99OCR** [comment]
Sorry, my wife had to go in for surgery two weeks ago, I've been away from the computer most of the time since then. 

Basically, the sorting order would be somewhat natural. Play next alphabetical folder in the list. Then play the files (sorted by filename alphanumeric, so track 01 - blah.vgm, then track 02 - whatevs.vgm). So if I had a folder of:

00 - Best Game
1 - Okay Game
07 - Revenge of Game
72 - Game the Sequels
A Game
Better than A Game
Doh! More Game!
Sorry, Not as Good as Game

It would play in that order.

As for the sorting problem with the playlists, I'll reproduce once I get back in front of my other machine.

**Comment by droidmjt** [comment]
Just click the colorful bar above the files and choose "sort by filename". Or go to settings and choose from Sorting preference -> Filename.

**Comment by Level99OCR** [comment]
Okay, I figured out my issue.

I am still getting the playlist naming issue where I enter one character, and the next character I enter always counts as a backspace. It alternates like that so I can only ever have playlists with a single character for the name. I've tried it on multiple devices and they all behave the same way.

As for the playlist sorting, it has to do with when I try to add a folder from within a ZIP/7Z archive. I'll highlight the three scenarios I've done:

Dreamcast -> [Alpha Subfolders] -> [ZIP Files]
In this case, I try to add the 'A' subfolder to playlist, and it works exactly how I would expect it to work. Each ZIP shows up as its own entry. However, once I start trying to play stuff and it gets to the end of the folder, it just loops back around to the beginning of the folder. So one game continues on loop forever.

DS -> [ZIP Files containing each Alpha grouping]
Here, I have a very large ZIP containing all the games that begin with A in their own folders, within  'A.ZIP' . I try to add the A zip to a playlist, and the options it presents me are simply rescan or delete.

[ZIP File containing all Alpha groupings of Genesis games]
Lastly, this is one big file called 'Genesis.ZIP' that contains all folders A-Z plus #, each of those containing the appropriate games in folders of their own name.  I try to add the A subfolder within the zip, and while it adds it, it adds all the files within it to the base of the playlist (the only option I have is to add all to playlist). So Altered Beast no longer is within its own subfolder, for example. All the games files are listed together in the base playlist and there's no way to sort them by game because the game name does not prefix each individual file.

I hope that makes sense. Thank you for keeping at this, it is greatly appreciated! As it is, I'm using it to make some temporary random playlists. It isn't without use but it still isn't quite in the configuration that I believe would be ideal. Whether or not the amount of coding to make it so would be too much is your call, as always :D

**Comment by droidmjt** [comment]
About the playlist naming thing, I have tried various images using emulator, 3 different samsung phones + Cyanogenmod 12.1 and i can't reproduce the problem.

The rest, yeah it makes sense somehow, I do some experiments and see how it goes, I post more questions here soon.

**Comment by droidmjt** [comment]
DS -> [ZIP Files containing each Alpha grouping]
Here, I have a very large ZIP containing all the games that begin with A in their own folders, within 'A.ZIP' . I try to add the A zip to a playlist, and the options it presents me are simply rescan or delete.

in above case, do you have zips inside zips? if yes, that is not supported and won't get supported

**Comment by Level99OCR** [comment]
No, I do not have ZIPs inside ZIPs. It's just one big zip for all the games that begin with the letter A. Since Droidsound E displays a ZIP file in its browser just like it would a regular directory in file hierarchy, the ZIPs name is A and all the A games are inside there in their own folders.

The devices I've tried the playlist thing on and they're showing the naming issue are:

ASUS Zenfone 2 (Lollipop 5.0, Stock Rom, Fresh Droidsound E install) Running R46
Moto G 4G LTE (Lollipop 5.1, Stock Rom, Fresh Droidsound E install) Running R52
Samsung Galaxy S3 AT&T (Lollipop 5.1, CM12.1, Dirty Droidsound E install) Running R44

**Comment by droidmjt** [comment]
wow, finally i was able to reproduce the playlist naming problem, will fix ASAP

**Comment by Level99OCR** [comment]
Take your time, man! Appreciated! Also, thank you VERY much for the volume modification and plugin enable/disable features!

**Comment by droidmjt** [comment]
I have solved the first problem by adding a manual queue option in context menu, click it anywhere over archive or dir or archivedir and it will queue everything under it to existing queue or a new one if there is no existing

**Comment by droidmjt** [comment]
https://www.dropbox.com/s/3f5flqfd0lggdz0/droidsounde1664_54.apk?dl=0

**Comment by Level99OCR** [comment]
Downloading shortly, thanks!

**Comment by droidmjt** [comment]
i just invented a better queuing method, i'll upload a new binary once its ready

**Comment by droidmjt** [comment]
just tried the new system with ~200000 songs in playqueue, no memory problems :)

**Comment by Level99OCR** [comment]
Oh awesome! I had been testing over the weekend, but I'll save my results for the latest binary. Thanks!

**Comment by droidmjt** [comment]
yeah, i'm still adjusting the queue stuff

**Comment by Level99OCR** [comment]
Tried on my Moto G dirty install and still having trouble making playlists with more than one letter. Doing a clean install now to see what's going on.

**Comment by sketch242** [comment]
I have been having that problem with playlist names for a few versions now.  I think the last time I created one I fixed it by renaming the plist on the phone.  Maybe we should file a separate bug for that.

**Comment by droidmjt** [comment]
shit, i thought i fixed it, i'll check it now again

**Comment by droidmjt** [comment]
in the meantime, disable predictive text and it should work

**Comment by droidmjt** [comment]
found the root cause for this problem, fixing...

**Comment by droidmjt** [comment]
https://www.dropbox.com/s/0jeyavu1qlwz63k/droidsounde1664_55.apk?dl=0

**Comment by droidmjt** [comment]
the above binary has the new queue system, new options available in context menu, let me know how it works

**Comment by Level99OCR** [comment]
Downloading and resting copiously over the weekend!

**Comment by Level99OCR** [comment]
*testing

**Comment by Level99OCR** [comment]
Playlist naming is definitely fixed. Trying to add a folder within a zip directory to playqueue causes forceclose. Still testing out permutations of the playqueue, will upload the crashlogs once the testing is done.

**Comment by droidmjt** [comment]
do you have build _55 ? I just tried to create new playqueue from Examples/Amiga and worked fine.

**Comment by Level99OCR** [comment]
I was using 55, but I'll try again. Been out of commission for the past few weeks, will have a status shortly.

**Comment by Level99OCR** [comment]
Did a clean install with 66. Play queue works but I can't add my external micro SD card as a file browser anymore. I just get the option for internal memory (seen as sd card) and no other file path.

**Comment by Level99OCR** [comment]
Just FYI this particular device mounts to /storage/MicroSD for its external storage path, but of course it varies with device.

**Comment by droidmjt** [comment]
what device it is? I'll see if i can get that fixed.

**Comment by Level99OCR** [comment]
It looks like the last version I have that allows me to add my external path is 1664_52. Somewhere between that and 1664_55 is where I lost the ability to add it. I'm guessing this, along with other more recent changes, could have affected my testing of the play queue.

**Comment by droidmjt** [comment]
no worries, i know the problem now, i'll fix it

**Comment by Level99OCR** [comment]
Zenfone 2 Z00A  I'm about to test with moto g, galaxy s3, lgog, and burst.

**Comment by Level99OCR** [comment]
Okay, thanks so much!

**Comment by droidmjt** [comment]
https://www.dropbox.com/s/9ywotabmvo8ktzv/droidsounde1664_67.apk?dl=0
try and let me know if it works

**Comment by Level99OCR** [comment]
That did it. Fantastic. Zenfone 2 is x86 running libhoudini to translate ARM code, so the adding of the songs again will take longer than normal, but I have started the scan and will report back on playqueue testing as soon as it completes. Hopefully can close this out soon.

**Comment by Level99OCR** [comment]
Works stupendously. Closing this, as feature is implemented. If at all possible in the future to make it able to view the current playqueue, that would be the only thing I feel is missing. Fantastic work!!!

**Comment by droidmjt** [comment]
yeah i've been thinking the same feature, to show the current queue somehow, the biggest problem is that, where exactly to show it, in what view?

x86 using libhoudini and translate stuff!?? crap, i always that its all native x86.

Sure, send the snes stuff, i'll see if i can get some more ideas of them.

**Comment by Level99OCR** [comment]
Will do. Also gonna email you about an x86 native build.
